File: courses/machine_learning/Regression/section_8_Polynomial_Regression/Python/polynomial_regression.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# Import data
dataset = pd.read_csv("Position_Salaries.csv")
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

# Training the Linear Regression model
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Training the Polynomial regression model
# y = b_0 + b_1(x_n)^n. Degree means how many power of 'n' we go through.
# Higher n = higher accuracy, which means overfitting which we do not want that
# if n = 1 it is linear regression model
poly_reg = PolynomialFeatures(degree=4)
# simple matrix of features containing (x_n)
X_poly = poly_reg.fit_transform(X)
# Training part
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

plt.figure(figsize=(12, 7))
# Visualising Linear Regression model
plt.subplot(1,2,1)
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg.predict(X), color="blue")
plt.title("Linear Regression model")
plt.xlabel("Position lvl")
plt.ylabel("Salary")

# Visualising Polynomial regression model
plt.subplot(1,2,2)
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg_2.predict(X_poly), color="blue")
plt.title("Polynomial Regression model")
plt.xlabel("Position lvl")
plt.ylabel("Salary")

# Prevent overlapping
plt.tight_layout()
plt.show()

# Predicting new results whit linear regression
print(lin_reg.predict([[6.5]]))

# Predicting new results whit polynomial regression
print(lin_reg_2.predict(poly_reg.fit_transform([[6.5]])))

File: courses/machine_learning/Regression/section_8_Polynomial_Regression/R/polynomial_regression.R
dataset = read.csv("Position_Salaries.csv")
dataset = dataset[2:3]

# Fitting linear regression to the dataset
# This no work sadge
lin_reg = lm(formula = Salary ~ Level,
             data = dataset)

# Fitting polynomial regression to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

library(ggplot2)
# Visualising Linear regression
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
                colour = "blue") +
  ggtitle("Linear Regression") +
  xlab("Level") +
  ylab("Salary")

# Visualising Polynomial regression
ggplot()+
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red")+
  geom_line(aes(x = dataset$Level, y = predict(poly_reg,newdata = dataset)),
                colour = "blue") +
  ggtitle("Polynomial Regression") +
  xlab("Level") +
  ylab("Salary")

# Predicting new result whit linear regression
predict(lin_reg, newdata = data.frame(Level = 6.5))

# Predicting new result whit Polynomial regression
predict(poly_reg, newdata = data.frame(Level = 6.5,
                                       Level2 = 6.5^2,
                                       Level3 = 6.5^3,
                                       Level4 = 6.5^4))